core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatEntry.scala

```diff
@@ -430,61 +438,85 @@ class ColumnFormatValue extends SerializedDiskBuffer
       assert(buffer.order() eq ByteOrder.LITTLE_ENDIAN)
-      val position = buffer.position()
-      val typeId = buffer.getInt(position)
+      position = buffer.position()
+      typeId = buffer.getInt(position)
       if (typeId >= 0) {
         this.decompressionState = 1
-        return this
-      }
-      // first check if decompression should be skipped
-      // (when onlyIfStored is true and underlying buffer cannot be replaced)
-      if (onlyIfStored) {
-        if (fromDisk || (isDirect && this.refCount > 2)) return this
-        // check if entry was read from disk without faultin
-        val entry = this.entry
-        if ((entry ne null) && entry.isValueNull) return this
+        false
+      } else {
+        context = this.regionContext
+        isDirect = buffer.isDirect
+        doReplace = !isDirect || this.refCount <= 2
+        updateStats = doReplace && (context ne null) && !fromDisk
+        // first check if decompression should be skipped
+        // (when onlyIfStored is true and underlying buffer cannot be replaced)
+        if (onlyIfStored) {
+          if (fromDisk || !doReplace) false
+          else {
+            // check if entry was read from disk without fault-in
+            val entry = this.entry
+            (entry eq null) || !entry.isValueNull
+          }
+        } else true
       }
-
+    })
+    // decompress and replace buffer if possible
+    if (!doDecompress) this
+    else {
       // replace underlying buffer if either no other thread is holding a reference
       // or if this is a heap buffer
       val allocator = GemFireCacheImpl.getCurrentBufferAllocator
-      val context = this.regionContext
       val perfStats = getCachePerfStats(context)
       val startDecompression = perfStats.startDecompression()
       val decompressed = CompressionUtils.codecDecompress(buffer, allocator, position, -typeId)
       try {
         // update decompression stats
         perfStats.endDecompression(startDecompression)
         val newValue = copy(decompressed, isCompressed = false, changeOwnerToStorage = false)
-        if (!isDirect || this.refCount <= 2) {
-          val updateStats = (context ne null) && !fromDisk
+        if (doReplace) {
+          val numBytes = decompressed.capacity() - buffer.capacity()
           if (updateStats && !isDirect) {
             // acquire the increased memory after decompression
-            val numBytes = decompressed.capacity() - buffer.capacity()
             if (!StoreCallbacksImpl.acquireStorageMemory(context.getFullPath,
               numBytes, buffer = null, offHeap = false, shouldEvict = true)) {
               throw LocalRegion.lowMemoryException(null, numBytes)
             }
           }
           val newBuffer = transferToStorage(decompressed, allocator)
-          // update the statistics before changing self
-          if (updateStats) {
-            context.updateMemoryStats(this, newValue)
+          val releaseStorageMemory = synchronized {
```

Comment:
Can transfertoStorage be also called for cases where we are actually replacing the buffer.

Reply:
transferToStorage is primarily for off-heap buffers that hold a Runnable inside which will release the storage in UMM and the buffer itself (either by GC or by explicit UnsafeHolder.releaseDirectBuffer). This Runnable inside the Cleaner in DirectByteBuffer will be changed to "FreeStoreBuffer".

The other change is just transforming buffer type if required, as per storage type: e.g. heap to off-heap if storage type is off-heap but incoming is heap. This case should normally not happen in our code paths.

So this call is for changing an existing buffer which will then be stuffed into ColumnFormatValue. For example, compression will use temporary memory (ideally execution memory but was not being done due to LME issues) which will then be transferred to storage accounting if compression has reduced the size enough else it is released right away.

Reply:
My question was a little different. I was not very clear. We are using "newBuffer" value only if certain conditions are met (like no other thread has changed "this.columnBuffer").  So why call transferToStorage if we ar enot going to use newBuffer i.e in the else part in line 512.

Reply:
That was done to keep transferToStorage outside the sync block because that can also call UMM. I guess this should also eagerly release newBuffer in case it was not stored in region so will add that.